File: back/src/main/java/com/triptrace/domain/image/image/entity/Image.java
package com.triptrace.domain.image.image.entity;

import com.triptrace.domain.member.member.entity.Member;
import com.triptrace.domain.post.post.entity.Post;
import com.triptrace.domain.trip.trip.entity.Trip;
import com.triptrace.global.jpa.entity.BaseEntity;
import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.EnumType;
import jakarta.persistence.Enumerated;
import jakarta.persistence.FetchType;
import jakarta.persistence.JoinColumn;
import jakarta.persistence.ManyToOne;
import java.math.BigDecimal;
import java.time.LocalDateTime;

@Entity
public class Image extends BaseEntity {
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "owner_id", nullable = false)
    private Member owner;
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "trip_id", nullable = false)
    private Trip trip;
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "post_id")
    private Post post;
    @Column(nullable = false)
    private String originalFileUrl;
    private String thumbnailUrl;
    @Column(nullable = false)
    private Long fileSize;
    @Column(length = 50, nullable = false)
    private String mimeType;
    // EXIF 분석 완료 후 채워지는 이미지 메타데이터
    // 정수 3자리, 소수 7자리
    @Column(precision = 10, scale = 7)
    private BigDecimal gpsLat;
    @Column(precision = 10, scale = 7)
    private BigDecimal gpsLng;
    private LocalDateTime capturedAt;
    private String deviceInfo;
    @Enumerated(EnumType.STRING)
    @Column(length = 10, nullable = false)
    private UploadStatus uploadStatus;

    //all args
    public Image(
        Member owner,
        Trip trip,
        Post post,
        String originalFileUrl,
        String thumbnailUrl,
        Long fileSize,
        String mimeType,
        BigDecimal gpsLat,
        BigDecimal gpsLng,
        LocalDateTime capturedAt,
        String deviceInfo,
        UploadStatus uploadStatus
    ){
        this.owner = owner;
        this.trip = trip;
        this.post = post;
        this.originalFileUrl = originalFileUrl;
        this.thumbnailUrl = thumbnailUrl;
        this.fileSize = fileSize;
        this.mimeType = mimeType;
        this.gpsLat = gpsLat;
        this.gpsLng = gpsLng;
        this.capturedAt = capturedAt;
        this.deviceInfo = deviceInfo;
        this.uploadStatus = uploadStatus;
    }

    // 업로드 시점에 알 수 있는 기본 이미지 정보만 먼저 저장
    public Image(
        Member owner,
        Trip trip,
        Post post,
        String originalFileUrl,
        String thumbnailUrl,
        Long fileSize,
        String mimeType,
        UploadStatus uploadStatus
    ) {
        this(owner,
            trip,
            post,
            originalFileUrl,
            thumbnailUrl,
            fileSize,
            mimeType,
            null,
            null,
            null,
            null,
            uploadStatus
        );
    }
    public void modifyPost(Post post){
        this.post = post;
    }

    public void disconnectPost() {
        this.post = null;
    }

    public void modifyStatus(UploadStatus uploadStatus) {
        this.uploadStatus = uploadStatus;
    }

    // 자동 생성 결과로 만들어진 Post에 이미지를 배치
    // 추후 이미지 관계 테이블 도입 시 관계 테이블로 이전
    public void connectPost(Post post) {
        this.post = post;
    }

    @java.lang.SuppressWarnings("all")
    public Member getOwner() {
        return this.owner;
    }

    @java.lang.SuppressWarnings("all")
    public Trip getTrip() {
        return this.trip;
    }

    @java.lang.SuppressWarnings("all")
    public Post getPost() {
        return this.post;
    }

    @java.lang.SuppressWarnings("all")
    public String getOriginalFileUrl() {
        return this.originalFileUrl;
    }

    @java.lang.SuppressWarnings("all")
    public String getThumbnailUrl() {
        return this.thumbnailUrl;
    }

    @java.lang.SuppressWarnings("all")
    public Long getFileSize() {
        return this.fileSize;
    }

    @java.lang.SuppressWarnings("all")
    public String getMimeType() {
        return this.mimeType;
    }

    @java.lang.SuppressWarnings("all")
    public BigDecimal getGpsLat() {
        return this.gpsLat;
    }

    @java.lang.SuppressWarnings("all")
    public BigDecimal getGpsLng() {
        return this.gpsLng;
    }

    @java.lang.SuppressWarnings("all")
    public LocalDateTime getCapturedAt() {
        return this.capturedAt;
    }

    @java.lang.SuppressWarnings("all")
    public String getDeviceInfo() {
        return this.deviceInfo;
    }

    @java.lang.SuppressWarnings("all")
    public UploadStatus getUploadStatus() {
        return this.uploadStatus;
    }

    @java.lang.SuppressWarnings("all")
    protected Image() {
    }
}

File: back/src/main/java/com/triptrace/domain/image/image/entity/Image.kt
package com.triptrace.domain.image.image.entity

import com.triptrace.domain.member.member.entity.Member
import com.triptrace.domain.post.post.entity.Post
import com.triptrace.domain.trip.trip.entity.Trip
import com.triptrace.global.jpa.entity.BaseEntity
import jakarta.persistence.*
import java.math.BigDecimal
import java.time.LocalDateTime

@Entity
class Image protected constructor() : BaseEntity() {
    @ManyToOne(fetch = FetchType.LAZY) @JoinColumn(name = "owner_id", nullable = false)
    lateinit var owner: Member
    @ManyToOne(fetch = FetchType.LAZY) @JoinColumn(name = "trip_id", nullable = false)
    lateinit var trip: Trip
    @ManyToOne(fetch = FetchType.LAZY) @JoinColumn(name = "post_id")
    var post: Post? = null
    @Column(nullable = false) lateinit var originalFileUrl: String
    var thumbnailUrl: String? = null
    @Column(nullable = false) var fileSize: Long = 0
    @Column(length = 50, nullable = false) lateinit var mimeType: String
    @Column(precision = 10, scale = 7) var gpsLat: BigDecimal? = null
    @Column(precision = 10, scale = 7) var gpsLng: BigDecimal? = null
    var capturedAt: LocalDateTime? = null
    var deviceInfo: String? = null
    @Enumerated(EnumType.STRING) @Column(length = 10, nullable = false)
    lateinit var uploadStatus: UploadStatus

    constructor(owner: Member, trip: Trip, post: Post?, originalFileUrl: String, thumbnailUrl: String?, fileSize: Long, mimeType: String, gpsLat: BigDecimal?, gpsLng: BigDecimal?, capturedAt: LocalDateTime?, deviceInfo: String?, uploadStatus: UploadStatus) : this() {
        this.owner = owner; this.trip = trip; this.post = post; this.originalFileUrl = originalFileUrl; this.thumbnailUrl = thumbnailUrl; this.fileSize = fileSize; this.mimeType = mimeType; this.gpsLat = gpsLat; this.gpsLng = gpsLng; this.capturedAt = capturedAt; this.deviceInfo = deviceInfo; this.uploadStatus = uploadStatus
    }
    constructor(owner: Member, trip: Trip, post: Post?, originalFileUrl: String, thumbnailUrl: String?, fileSize: Long, mimeType: String, uploadStatus: UploadStatus) : this(owner, trip, post, originalFileUrl, thumbnailUrl, fileSize, mimeType, null, null, null, null, uploadStatus)

    fun modifyPost(post: Post?) { this.post = post }
    fun disconnectPost() { post = null }
    fun modifyStatus(uploadStatus: UploadStatus) { this.uploadStatus = uploadStatus }
    fun connectPost(post: Post) { this.post = post }
}
